xhistogram/core.py

```diff
@@ -128,7 +128,23 @@ def _histogram_2d_vectorized(*args, bins=None, weights=None, density=False,
     # just throw out everything outside of the bins, as np.histogram does
     # TODO: make this optional?
     slices = (slice(None),) + (N_inputs * (slice(1, -1),))
-    return bin_counts[slices]
+    bin_counts = bin_counts[slices]
+
+    if density:
+        bin_widths = [np.diff(b) for b in bins]
+        if N_inputs == 1:
+            bin_areas = bin_widths[0]
+        elif N_inputs == 2:
+            bin_areas = np.outer(*bin_widths)
+        else:
+            # TODO use np.einsum for N-D case?
+            raise NotImplementedError("density=True not implemented for "
+                                      "histograms of dimension > 2, but there "
+                                      f"are {N_inputs} input variables")
```

Comment:
I think we should really try to handle ND.  It should not be too hard. This logic looks right, we just need to generalize it. Perhaps an easier path would be to use xarray datasets for the widths, which will automatically handle the outer product. But we don't want an xarray dependency in core...

Reply:
Also, are we sure we need this function *inside* this loop. Or could we do the normalization at a higher level, like inside `histogram`.

Reply:
I agree that this could be moved to `histogram`

Then, what about something like:
```python
if density:
    bin_widths = [np.diff(b) for b in bins]
    if N_inputs == 1:
        bin_areas = bin_widths[0]
    else:
        bin_areas = np.prod(np.ix_(*bin_widths))
    h = h / bin_areas / h.sum()
```
You could also enforce that `N_inputs==2` uses `np.outer(*bin_widths)`, which should be faster than `np.prod(np.ix_(*bin_widths))`.

Note that moving the density logic to `histogram` also requires that `_ensure_bins_is_a_list_of_arrays(bins, N_inputs)` is moved there, but I don't see why this should be an issue